In [1]:
import tensorflow as tf

In [2]:
# building a computational graph
tf.compat.v1.disable_eager_execution()
# defining constant
node1 = tf.constant(3.0, tf.float32)
node2 = tf.constant(4.0)


In [3]:
# We have to first create session in order to run tensorflow 
sess = tf.compat.v1.Session()
print(sess.run([node1, node2]))
# We have to close session once our task is completed
sess.close()

[3.0, 4.0]


In [4]:
# have to disable eager execution in TF2.0
tf.compat.v1.disable_eager_execution()
# Adding 2 numbers using tensorflow
a = tf.constant(2)
b = tf.constant(5)

d = tf.add(a,b)

# Creating Session
sess = tf.compat.v1.Session()
g = sess.run(d)
print(g)
sess.close()

7


 - By constant we have to provide the value in starting.
 - to overcome this we use placeholder by which we can provide values later

In [5]:
tf.compat.v1.disable_eager_execution()

# first we have to provide type of variable first
a = tf.compat.v1.placeholder(tf.float32)
b = tf.compat.v1.placeholder(tf.float32)

c = a*b

sess = tf.compat.v1.Session()
# Providing values to the placeholder
sess.run(c,{a:[1,2], b:[2,3]})
sess.close()

In order to make our model trainable we have to introduce Variables that allow use to as trainable parameters to our graph

In [6]:
tf.compat.v1.disable_eager_execution()

# Initilizing variables
W = tf.Variable([.3], tf.float32)
b = tf.Variable([-.3], tf.float32)
x = tf.compat.v1.placeholder(tf.float32)

c = W*x + b

# initlizing our globale variables
init = tf.compat.v1.global_variables_initializer()

# starting session
sess = tf.compat.v1.Session()
sess.run(init)

# Proving value of our placeholder
print(sess.run(c,{x:[2,4,6,5]}))

Instructions for updating:
If using Keras pass *_constraint arguments to layers.
[0.3        0.90000004 1.5        1.2       ]


### Loss Function

- A loss function measures how far our predicted value is from actual value

In [15]:
# To evaluate the model on training data
y = tf.compat.v1.placeholder(tf.float32)

squared_deltas = tf.compat.v1.square(c - y)

loss = tf.compat.v1.reduce_sum(squared_deltas)

print(sess.run(loss, {x:[1,2,3,4], y:[0,-1,-2,-3]}))

23.66


As we can see our loss is much more higher. In order to reduce our loss Tensorflow provide Gradient Descent

In [17]:
# Creating optimizer to reduce our loss
# 0.01 will be our learning rate
optimizer = tf.compat.v1.train.GradientDescentOptimizer(0.01)

train = optimizer.minimize(loss)

sess.run(init)
for i in range(1000):
    sess.run(train, {x:[1,2,3,4], y:[0,-1,-2,-3]})
print(sess.run([a,b]))

[array([-0.9999969], dtype=float32), array([0.9999908], dtype=float32)]


To minimize loss value for a should be -0.9999969 and b should be 0.9999908